lib/OpenQA/Worker/Cache.pm

```diff
@@ -263,106 +236,46 @@ sub get_asset {
-        $res = $dbh->prepare($sql)->execute($toggle, $asset, $asset, $check);
-        $dbh->commit or die $dbh->errstr;
-    };
-
-    if ($@) {
-        log_error "toggle_asset_lock: Rolling back $@";
-        eval { $dbh->rollback };
-        log_error "Rolling back failed: $@" if $@;
-    }
-
-    return 0 if !defined $res || $res == 0 || $@;
-    return 1 if defined $res && $res == 1;
-}
-
 sub _asset {
     my ($self, $asset) = @_;
+    my $result = $self->dbh->db->select('assets', [qw(etag size last_use)], {filename => $asset})->hashes;
 
-    my $sql = "SELECT downloading, etag from assets where filename = ?";
-    $self->dbh->prepare($sql);
-    my $result = $self->dbh->selectrow_hashref($sql, undef, $asset);
-
-    return {} unless $result;
-    return $result;
-}
-
-sub try_lock_asset {
-    my ($self, $asset) = @_;
-    my $sth;
-    my $sql;
-    my $lock_granted = 0;
-    my $result;
-
-    my $dbh = $self->dbh;
-
-    eval {
-        $lock_granted = 1 if $self->toggle_asset_lock($asset, 1);
-        $result = $self->_asset($asset);
-        log_info "CACHE: Being downloaded by another worker, sleeping."
-          if exists $result->{downloading} && $result->{downloading} == 1 && $lock_granted == 0;
-    };
-    log_error "try_lock_asset: Failed $@" if $@;
-
-    return $result if $lock_granted;
-    return !!0;
+    return {} if $result->size == 0 || $@;
+    return $result->first;
 }
 
 sub track_asset {
     my ($self, $asset) = @_;
 
-    my $dbh = $self->dbh;
     my $res;
-    my $sql
-      = "INSERT OR IGNORE INTO assets (downloading,filename, size, last_use) VALUES (0, ?, 0,  strftime('%s','now'));";
+    my $sql = "INSERT OR IGNORE INTO assets (filename, size, last_use) VALUES (?, 0,  strftime('%s','now'));";
 
     eval {
-        $dbh->do("BEGIN EXCLUSIVE");
-        $res = $dbh->prepare($sql)->execute($asset) or die $dbh->errstr;
-        $dbh->commit or die $dbh->errstr;
+        my $tx = $self->dbh->db->begin('exclusive');
+        $res = $self->dbh->db->query($sql, $asset)->arrays;
+        $tx->commit;
     };
-
     if ($@) {
         log_error "track_asset: Failed: $@";
-        eval { $dbh->rollback };
-        log_error "Rolling back failed: $@";
     }
 
-    return !!0 if !defined $res || $res == 0 || $@;
-    return !!1 if defined $res;
+    return !!0 if $res->size == 0 || $@;
+    return !!1 if $res->size > 0;
 }
 
 sub _update_asset_last_use {
     my ($self, $asset) = @_;
 
-    my $dbh = $self->dbh;
-    $dbh->do("BEGIN EXCLUSIVE");
-
-    my $sql = "UPDATE assets set last_use = strftime('%s','now') where filename = ?;";
     eval {
-        my $sth = $dbh->prepare($sql);
-        $sth->bind_param(1, $asset);
-        $sth->execute;
-        $dbh->commit;
+        my $tx  = $self->dbh->db->begin('exclusive');
+        my $sql = q(UPDATE assets set last_use = strftime('%s','now') where filename = ?;);
+        $tx->commit;
+        $self->dbh->db->query($sql, $asset);
+        $tx->commit;
```

Comment:
That this works is pure luck. You need to hold on to the db handle for the duration of the transaction. The call to `$self->dbh->db` can give you a random handle from the pool if there's more than one connection active. It might not be the one the transaction is active on.

Reply:
Thanks for looking at it :-) I misunderstood the docs ( https://metacpan.org/pod/Mojo::SQLite::Database#begin ), I had the (wrong) impression that the transaction was holding that. 
Should be fixed in https://github.com/os-autoinst/openQA/pull/1783/commits/6ee6a91d33bf22669034654d05645a5d89ac71ab